Source/Core/Monomorphization.cs

```diff
@@ -1378,13 +1392,123 @@ public static MonomorphizationVisitor Initialize(CoreOptions options, Program pr
         monomorphizationVisitor.polymorphicMapInfos.Values.Select(polymorphicMapInfo =>
           polymorphicMapInfo.CreateDatatypeTypeCtorDecl(polymorphicMapAndBinderSubstituter)).ToList();
       var splitAxioms = polymorphicMapAndBinderSubstituter.Substitute(program);
+      UpdateDependencies(monomorphizationVisitor);
       program.RemoveTopLevelDeclarations(decl => decl is Axiom);
       program.AddTopLevelDeclarations(splitAxioms);
       program.AddTopLevelDeclarations(polymorphicMapDatatypeCtorDecls);
       Contract.Assert(MonomorphismChecker.IsMonomorphic(program));
       return monomorphizationVisitor;
     }
 
+    private static void UpdateDependencies(MonomorphizationVisitor monomorphizationVisitor)
+    {
+      // Original program axioms were replaced with new splitAxioms. Program constants and functions that had
+      // those original axioms as dependencies need their references updated to the new axioms.
+      // Axioms / functions could both have been instantiated, which adds some complexity.
+
+      foreach (var (oldAxiom, newAxioms) in monomorphizationVisitor.originalAxiomToSplitAxioms)
+      {
+        var oldAxiomFC = new FunctionAndConstantCollector();
+        var unmodifiedOldAxiom = monomorphizationVisitor.originalAxiomToUninstantiatedCopies[oldAxiom];
+        oldAxiomFC.Visit(unmodifiedOldAxiom);
+        var oldAxiomFunctions = oldAxiomFC.Functions;
+
+        Dictionary<Axiom, HashSet<Function>> newAxiomFunctions = new Dictionary<Axiom, HashSet<Function>>();
+        Dictionary<Axiom, HashSet<Constant>> newAxiomConstants = new Dictionary<Axiom, HashSet<Constant>>();
+
+        foreach (var newAxiom in newAxioms)
+        {
+          var newAxiomFC = new FunctionAndConstantCollector();
+          newAxiomFC.Visit(newAxiom);
+          newAxiomFunctions.Add(newAxiom, newAxiomFC.Functions);
+          newAxiomConstants.Add(newAxiom, newAxiomFC.Constants);
+        }
+
+        void UpdateFunctionDependencies(HashSet<Function> functions)
```

Comment:
Adding the parameter `HashSet<Function> functions` seems unnecessary, since it's also available through `monomorphizationVisitor.originalFunctions`

I'm not always sure how much inner functions help read the code. Here I have to inspect the outer scope to understand the meaning of some of the variables used.